tests/unittests/tests-uri_parser/tests-uri_parser.c

```diff
@@ -21,599 +21,601 @@
-        "",
-        "",
-        "",
-        "",
-        0,
-        "./this:that",
-        "",
-        0),
-    VEC("pP://",
-        true,
-        "pP",
-        "",
-        "",
-        "",
-        "",
-        "",
-        0,
-        "",
-        "",
-        0),
-    VEC("A://@",
-        true,
-        "A",
-        "",
-        "",
-        "",
-        "",
-        "",
-        0,
-        "",
-        "",
-        0),
-    VEC("coap://example.com:1234568910",
-        /* is URI */
-        true,
-        /* parsed scheme */
-        "coap",
-        /* parsed userinfo */
-        "",
-        /* parsed host */
-        "example.com",
-        /* parsed host without zoneid */
-        "",
-        /* parsed zoneid */
-        "",
-        /* parsed port_str */
-        "",
-        /* parsed port */
-        0,
-        /* parsed path */
-        "",
-        /* parsed query */
-        "",
-        /* expected return value */
-        -1),
-    VEC("coap://example.com: 12/foo",
-        true,
-        "",
-        "",
-        "example.com",
-        "",
-        "",
-        "",
-        0,
-        "",
-        "",
-        -1),
-    VEC("coap://example.com:0x12/foo",
-        true,
-        "",
-        "",
-        "example.com",
-        "",
-        "",
-        "",
-        0,
-        "",
-        "",
-        -1),
-};
-
-static char _failure_msg[VEC_MSG_LEN];
-
-static void test_uri_parser__validate(void)
+ * The uri_parser returns string-buffer(not zero terminated strings).
+ * This function compares a given string with a given uri_parser buffer + length.
+ * Additionally, it takes the original uri (from the test vector) and a name
+ * in order to print a precise and helpful error message if the buffer and string
+ * aren't identical.
+ */
+static int _compare_string_buffer(const char *name, const char *input_uri, const char *expected_str,
+                                  const char *actual_str, size_t actual_len)
 {
-    uri_parser_result_t ures;
-    for (unsigned i = 0; i < ARRAY_SIZE(validate_uris); i++) {
-        int res = uri_parser_process_string(&ures, validate_uris[i].uri);
-        TEST_ASSERT_EQUAL_INT(validate_uris[i].full_uri,
-               uri_parser_is_absolute_string(validate_uris[i].uri));
-        TEST_ASSERT_EQUAL_INT(validate_uris[i].expected, res);
-        if (res == 0) {
-            VEC_CHECK_STR(scheme, i, _failure_msg);
-            VEC_CHECK_STR(userinfo, i, _failure_msg);
-            VEC_CHECK_STR(host, i, _failure_msg);
-            VEC_CHECK_STR(ipv6addr, i, _failure_msg);
-            VEC_CHECK_STR(zoneid, i, _failure_msg);
-            VEC_CHECK_STR(port_str, i, _failure_msg);
-            VEC_CHECK_INT(port, i, _failure_msg);
-            VEC_CHECK_STR(path, i, _failure_msg);
-            VEC_CHECK_STR(query, i, _failure_msg);
+    if ((actual_len == strlen(expected_str)) &&
+        (strncmp(actual_str, expected_str, actual_len) == 0)) {
+        return 0;
+    }
+    printf(
```

Comment:
Just out of curiosity, you have actually looked at the output of this, right?

Reply:
Yes ofc. 
It looks like this: 
```
main(): This is RIOT! (Version: 2018.01-devel-26059-g6c24d-tests/uri_parser)
Help: Press s to start test, r to print it is ready
s
START
......
URI-Input: co ap://example.org/
Expected return: -1
Observed return: 0
uri_parser_tests._error_if_scheme_is_invalid (tests/unittests/tests-uri_parser/tests-uri_parser.c 371) Failure: The uri_parser failed to reject an invalid scheme.
.....
run 11 failures 1
```